solidity/contracts/bridge/Bridge.sol

```diff
@@ -794,10 +949,574 @@ contract Bridge is Ownable {
+            // the main UTXO and allow the wallet redeem their faults by
+            // transferring their outstanding funds to another healthy wallet.
+            // Note the redemption key in that case is:
+            // keccak256(walletPubKeyHash | walletPubKeyHash) since the
+            // `walletPubKeyHash` takes place of the usual `redeemerOutputHash`.
+            require(
+                invalidRedemptions[
+                    uint256(
+                        keccak256(
+                            abi.encodePacked(walletPubKeyHash, walletPubKeyHash)
+                        )
+                    )
+                ],
+                "Redemption transaction contains single change output"
+            );
+        }
+
+        return info;
+    }
+
+    // TODO: Function `notifyRedemptionTimeout. That function must:
+    //       1. Take a the `walletPubKey` and `redeemerOutputHash` as params.
+    //       2. Build the redemption key using those params.
+    //       3. Use the redemption key and take the request from
+    //          `pendingRedemptions` mapping.
+    //       4. If request doesn't exist in mapping - revert.
+    //       5. If request exits, and is timed out - remove the redemption key
+    //          from `pendingRedemptions` and put it to `invalidRedemptions`.
+    //          No need to check if `invalidRedemptions` mapping already contains
+    //          that key because `requestRedemption` blocks requests targeting
+    //          non-active wallets so there is no possibility that the given
+    //          redemption key could be reported as timed out multiple times.
+    //          On the other hand, if given redemption key was already
+    //          marked as invalid due to an amount-related fraud, it will not
+    //          be possible to report a time out on it since it won't be
+    //          present in `pendingRedemptions` mapping.
+    //       6. Return the `requestedAmount` to the `redeemer`.
+    //       7. Reduce the `pendingRedemptionsValue` (`wallets` mapping) for
+    //          given wallet by request's redeemable amount computed as
+    //          `requestedAmount - treasuryFee`.
+    //       8. Punish the wallet, probably by slashing its operators.
+    //       9. Change wallet's state in `wallets` mapping to `MovingFunds` in
+    //          order to prevent against new redemption requests hitting
+    //          that wallet.
+    //      10. Expect the wallet to transfer its funds to another healthy
+    //          wallet (just as in case of failed heartbeat).
+
+    // TODO: Function `submitRedemptionFraudProof`. That function must:
+    //       1. Take a `BitcoinTx.Info` and `BitcoinTx.Proof` of the
+    //          fraudulent transaction. It should also accept `walletPubKeyHash`.
+    //       2. Perform SPV proof to make sure it occurred on Bitcoin chain.
+    //          If not - revert.
```

Comment:
I can not recall - we are doing SPV proof because we can't validate the signature over the transaction? Can you please remind me?

Reply:
SPV is simpler in that context. Checking only the signature is tricky since we must extract the potentially fraudulent outputs from the sighash which is far more complicated than doing it using SPV-proven transaction data. Although, this is still an open topic and the TODOs about those functions are here to not lose important details which arose during designing the redemption mechanism.